From e24521169eb648ffeeebea750a4987ccf894ae71 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Tue, 8 Sep 2020 15:51:22 +0200
Subject: [PATCH 11/12] zynqmp-topic-miamiplusmp.dts: Support hardware revision
 v1r1

Add a GPIO hog for the ethernet clock switch
Replaced pca9554a by pcal9554b (latched interrupt and configurable IO, address 0x20)
Enable PHY interrupt through MIO27
Enable 125 MHz recovered clock from ethernet PHY
Simpler Si5345 clock tree using 100MHz USB3 reference clock
---
 .../dts/xilinx/zynqmp-topic-miamiplusmp.dts   | 54 ++++++++++++++-----
 1 file changed, 40 insertions(+), 14 deletions(-)

diff --git a/arch/arm64/boot/dts/xilinx/zynqmp-topic-miamiplusmp.dts b/arch/arm64/boot/dts/xilinx/zynqmp-topic-miamiplusmp.dts
index 418797ace8a8..190960f9e8c1 100644
--- a/arch/arm64/boot/dts/xilinx/zynqmp-topic-miamiplusmp.dts
+++ b/arch/arm64/boot/dts/xilinx/zynqmp-topic-miamiplusmp.dts
@@ -21,7 +21,7 @@
 #include <dt-bindings/phy/phy.h>
 
 / {
-	model = "Topic Miami MPSoC";
+	model = "Topic Miami Plus MPSoC";
 	compatible = "topic,miamiplusmp", "xlnx,zynqmp";
 
 	aliases {
@@ -112,13 +112,19 @@
 	clock-frequency = <400000>;
 
 	/* GPIO expander */
-	gpioex: pca9554@38 {
-		compatible = "nxp,pca9554";
+	gpioex: pcal9554@20 {
+		compatible = "nxp,pcal9554b";
 		vcc-supply = <&reg_1v8_miami>;
-		reg = <0x38>; /* PCA9554A adress starts at 0b111000 */
-		gpio-line-names = "USERLED", "", "ETH_RESET_N", "USB0_RESET_N", "USB1_RESET_N", "", "", "";
+		reg = <0x20>;
+		gpio-line-names = "USERLED", "", "ETH_RESET_N", "USB0_RESET_N", "USB1_RESET_N", "CLK_SW_CNTRL", "", "";
 		gpio-controller;
 		#gpio-cells = <2>;
+		eth-clk-sw-ctrl {
+			gpio-hog;
+			gpios = <5 0>;
+			output-low;
+			line-name = "eth_clk_sw_ctrl";
+		};
 	};
 
 	/* EEPROM */
@@ -149,6 +155,11 @@
 		silabs,pll-m-num = <280>; /* PLL at 280x50M = 14.0 GHz */
 		silabs,pll-m-den = <1>;
 
+		/*
+		 * Clock tree for the ttpzu9 and tdpzu9, with one PLL running at
+		 * 1000MHz we can create 25, 100, 125 and 250 using even
+		 * dividers.
+		 */
 		assigned-clocks = <&si5345 1 0>,
 				  <&si5345 1 1>,
 				  <&si5345 1 2>,
@@ -167,17 +178,17 @@
 		assigned-clock-parents = <0>, <0>, <0>, <0>, <0>,
 					<&si5345 1 0>, /* out 0 */
 					<&si5345 1 0>,
-					<&si5345 1 2>,
 					<&si5345 1 0>,
 					<&si5345 1 0>,
-					<&si5345 1 2>, /* out 5 */
+					<&si5345 1 0>,
+					<&si5345 1 0>, /* out 5 */
+					<&si5345 1 0>,
 					<&si5345 1 0>,
 					<&si5345 1 0>,
-					<&si5345 1 1>,
 					<&si5345 1 0>;
-		assigned-clock-rates =	<400000000>, /* synth 0 */
-					<104000000>,
-					<250000000>,
+		assigned-clock-rates =	<1000000000>, /* synth 0 */
+					<0>,
+					<0>,
 					<0>,
 					<0>,
 					<100000000>, /* out 0 */
@@ -188,16 +199,18 @@
 					< 25000000>, /* out 5 (ethernet) */
 					<100000000>,
 					<100000000>,
-					< 26000000>, /* out 8 (USB) */
+					<100000000>, /* out 8 (PS refclk3) */
 					<100000000>;
 
 		out@0 {
+			/* MGT_D_REFCLK_0 */
 			reg = <0>;
 			silabs,format = <1>; /* LVDS 3v3 */
 			silabs,common-mode = <3>;
 			silabs,amplitude = <3>;
 		};
 		out@1 {
+			/* MGT_D_REFCLK_2 */
 			reg = <1>;
 			silabs,format = <1>; /* LVDS 3v3 */
 			silabs,common-mode = <3>;
@@ -220,6 +233,7 @@
 			always-on;
 		};
 		out@4 {
+			/* MGT_D_REFCLK_6 */
 			reg = <4>;
 			silabs,format = <1>; /* LVDS 3v3 */
 			silabs,common-mode = <3>;
@@ -234,6 +248,7 @@
 			always-on;
 		};
 		out@6 {
+			/* MGT_D_REFCLK_4 */
 			reg = <6>;
 			silabs,format = <1>; /* LVDS 3v3 */
 			silabs,common-mode = <3>;
@@ -241,6 +256,7 @@
 		};
 
 		out@7 {
+			/* CLOCK_FPGA0 */
 			reg = <7>;
 			silabs,format = <1>; /* LVDS 1v8 */
 			silabs,common-mode = <13>;
@@ -257,6 +273,7 @@
 		};
 
 		out@9 {
+			/* CLOCK_FPGA1 */
 			reg = <9>;
 			silabs,format = <1>; /* LVDS 1v8 */
 			silabs,common-mode = <13>;
@@ -291,12 +308,12 @@
 		};
 		partition@qspi-rootfs {
 			label = "qspi-rootfs";
-			reg = <0x180000 0x7e80000>;
+			reg = <0x180000 0xFE80000>;
 		};
 		/* Everything */
 		partition@qspi-all {
 			label = "qspi-all";
-			reg = <0x0 0x8000000>;
+			reg = <0x0 0x10000000>;
 		};
 	};
 };
@@ -368,6 +385,15 @@
 			device_type = "ethernet-phy";
 			reg = <0x1>;
 			reset-gpios = <&gpioex 2 1>; /* Active low, on port 2 of GPIO expander */
+			interrupt-parent = <&gpio>;
+			interrupts = <27 IRQ_TYPE_LEVEL_LOW>;
+			marvell,reg-init =
+				/* Output 125MHz sync-E clock */
+				<2 0x10 0xf1fc 0x0e01>, /* Reg 2,16 enable RCLK */
+				/* LED[0:2] irq (hi-z), MODE4 */
+				<3 0x10 0 0x061f>,
+				/* Adjust LED drive to 50% */
+				<3 0x11 0 0x4400>;
 		};
 	};
 };
-- 
2.17.1

